aptos-move/move-examples/staking/sources/commission.move

```diff
@@ -0,0 +1,145 @@
+        operator: address,
+        /// The yearly commission rate in dollars. Will be used to determine how much APT the operator receives.
+        yearly_commission_amount: u64,
+        /// Used to withdraw commission.
+        signer_cap: SignerCapability,
+        /// Timestamp for tracking yearly commission.
+        last_update_secs: u64,
+        /// Amount of debt in dollars owed to the operator due to insufficient amount received from node commission.
+        /// This can happen if the commission rate is set too high or APT price is too low.
+        commission_debt: u64,
+    }
+
+    fun init_module(commission_signer: &signer) {
+        let signer_cap = resource_account::retrieve_resource_account_cap(commission_signer, @manager);
+        move_to(commission_signer, CommissionConfig {
+            manager: @manager,
+            operator: @operator,
+            yearly_commission_amount: INITIAL_COMMISSION_AMOUNT,
+            signer_cap,
+            last_update_secs: timestamp::now_seconds(),
+            commission_debt: 0,
+        });
+    }
+
+    #[view]
+    public fun operator(): address acquires CommissionConfig {
+        CommissionConfig[@staking].operator
+    }
+
+    #[view]
+    public fun yearly_commission_amount(): u64 acquires CommissionConfig {
+        CommissionConfig[@staking].yearly_commission_amount
+    }
+
+    #[view]
+    public fun commission_owed(): u64 acquires CommissionConfig {
+        let config = &CommissionConfig[@staking];
+        // Commission earned so far = per second commission rate * seconds passed.
+        let now_secs = timestamp::now_seconds();
+        let seconds_passed = now_secs - config.last_update_secs;
+        let commission_earned = math64::mul_div(seconds_passed, config.yearly_commission_amount, ONE_YEAR_IN_SECONDS);
+
+        commission_earned + config.commission_debt
+    }
+
+    /// Can be called by the manager to change the yearly commission amount.
+    public entry fun set_yearly_commission_amount(manager: &signer, new_commission: u64) acquires CommissionConfig {
+        assert_manager(manager);
+        CommissionConfig[@staking].yearly_commission_amount = new_commission;
+    }
+
+    /// Can be called by the manager or operator to change the account that receives the commission.
+    public entry fun set_operator(account: &signer, new_operator: address) acquires CommissionConfig {
+        assert_manager_or_operator(account);
+        CommissionConfig[@staking].operator = new_operator;
+    }
+
+    /// Distribute the commission to operator and remaining amount to manager.
+    /// Can only be called by the manager or operator.
+    public entry fun distribute_commission(account: &signer) acquires CommissionConfig {
+        assert_manager_or_operator(account);
+
+        // Commission owed so far plus any debt.
+        let commission_owed = commission_owed() as u128;
```

Comment:
you have to consider the fractional value. Otherwise, in extreme cases, the commission can be all zero. For example, in your default setup, if this function is called every 300s, `commission_owed` = 0 always, and you reset the `last_update_secs` every time. So for the whole year, the accumulated commission is 0 but actually it's 100000.

you need something similar to `fractional_accumulated` we put in the following code to take fractional commission into consideration and refactor this function plus `commission_owed()` I guess?
https://github.com/aptos-labs/aptos-core/blob/21c64ae8293d49eade2be7dc7af2863291b6ab33/aptos-move/framework/aptos-framework/sources/account/rate_limiter.move#L27

Reply:
This is in octas already with 8 decimals. There will be some rounding for sure but at most 1 unit lost which likely wouldn't be a big loss

Reply:
I think this issue is not resolved. your `commission_owed` will always be zero if called every 300s if yearly_commission is 100000 usd.

Reply:
There's an assumption here that this will only used for node operator commission (as stated in the description as well) so you can't call this that often as the balance will be 0 until commission is paid out

Reply:
I also added a min balance check anyway to prevent distribution from being called when there's not enough balance. This would minimize impact from rounding errors (doesn't mitigate fully though)